### ISSUE TITLE
[Bug?]: The example in the docs doesn't seam to work.

### ISSUE THREAD

**Issue by RonenEizen** [solidjs/solid-start#1466]
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Current behavior 😯

The example in the [documentation](https://start.solidjs.com/api/RequestEvent) doesn't seam to work.

```
declare module "@solidjs/start/server" {
  interface RequestEventLocals {
    myNumber: number;
    someString: string;
  }
}
```

`locals` property within the RequestEvent doesn't have any types declared in global.d.ts.  For example in  one of the routes:

```
import { getRequestEvent } from 'solid-js/web'

const getMessage = cache(async () => {
  'use server'
  const event = getRequestEvent()
  // event?.locals is not typed
  return
}, 'message')

export const route = {
  load: () => getMessage(),
}
```

and my global.d.ts has

```
declare module '@solidjs/start/server' {
  interface RequestEventLocals {
    myNumber: number
    someString: string
  }
}
```

### Steps to reproduce 🕹

Steps:

1. Create an solid-start app using one of the templates
2. Declare  '@solidjs/start/server'  within the global.d.ts file
3. Try accessing `event.locals.<SOMETHING>` within a route function

**Comment by ryansolid** [comment]
I am going to ask for a reproduction. We have an example template `experiments` that has this usage in it and it appears to be working.

https://github.com/solidjs/solid-start/blob/main/examples/experiments/src/entry-server.tsx#L4-L9
https://github.com/solidjs/solid-start/blob/main/examples/experiments/src/routes/index.tsx#L12-L13

(the example uses an undeclared value, but switch it to n/s shows the right types).

**Comment by RonenEizen** [comment]
@ryansolid It does work if declared within entry-server.tsx. I created a [repository](https://github.com/RonenEizen/solid-typescript-bug-example/blob/main/src/routes/index.tsx) based on a basic template. The types still don't work when declared within global.d.ts file. Look at the line 9 within the `routes/index.tsx` file.

**Comment by ryansolid** [comment]
Oh.. hmm so this is a weird typescript-ism.. Lovely. Not sure what to do with this then.

**Comment by RonenEizen** [comment]
I am personally fine with the workaround. Should I keep this issue? Or would you rather have it resolved?

**Comment by AirBorne04** [comment]
@ryansolid i guess it would make sense to rework to namespace? i threw together the needed changes here #1469  -> than the new way to overwrite the `RequestEventLocals` would look like this:
```
/// <reference types="@solidjs/start/env" />

declare namespace App {
  interface RequestEventLocals {
    myNumber: number;
    someString: string;
  }
}
```

**Comment by rmacfie** [comment]
I think the only thing needed to get the `@solidjs/start/env` types working is to include `./env.d.ts` in the `exports` property of `package.json`.

The schema documentation for package.json says:

> The "exports" field is used to restrict external access to non-exported module files, also enables a module to import itself using "name".

**Comment by AirBorne04** [comment]
@rmacfie I did a couple of tests to figure it out and I did not test the specific suggestion. Did you test this?

For my understanding the problem has nothing to do with the actual reference vs. import. Just TypeScript switching the module interpretation from `ambient` (with reference) to `augmented` with any import / export statement, no need to import the `env.d.ts` file at all.

**Comment by rmacfie** [comment]
> @rmacfie I did a couple of tests to figure it out and I did not test the specific suggestion. Did you test this?

Yes, I modified `node_modules/@solidjs/start/package.json` to include `./env.d.ts` in the `exports` section. This makes references to `@solidjs/start/env` work.

If I understand correctly, once `exports` in `package.json` is declared it will act as a white-list for Typescript, so it's logical that `@solidjs/start/env` just doesn't work correctly until this is fixed.

**Comment by AirBorne04** [comment]
@rmacfie can you share the test setup, i tried making the stated modifications but it does not work for me.

**Comment by rmacfie** [comment]
> I think I got to the bottom of the behaviour, it is described [here ambient vs augmented modules](https://www.typescriptlang.org/docs/handbook/modules/reference.html#ambient-modules). Just adding an export {} is fixing the issue but seems weird to just have it dangling there with no meaning.

It may look weird but I believe this is the recommended pattern to follow for this situation.

It has probably worked at some point without `export {}`, but a change in `tsconfig.json` broke it, possible the addition of `"isolatedModules": true`.

**Comment by AirBorne04** [comment]
@rmacfie 

> > @rmacfie can you share the test setup, i tried making the stated modifications but it does not work for me.
> 
> I made a proof of concept of changes in solid-start and the basic example with some dummy extensions to the `ImportMetaEnv` and `RequestEventLocals` types. You can see the diff here: [main...rmacfie:solid-start:issue-1466-poc](https://github.com/solidjs/solid-start/compare/main...rmacfie:solid-start:issue-1466-poc)
> 
> In `examples/basic/src/routes/about.tsx` you can see
> 
> ```ts
>   console.log(import.meta.env.FOO);
>   console.log(requestEvent.locals.foo);
> ```
> 
> The foo properties are typed according to the extensions declared in `examples/basic/src/global.d.ts`.

Yes that is working because adding an empty `export {};` in  `examples/basic/src/global.d.ts` is telling typescript to interpret as an module augmentation. No other change needed from my testing experience.

> > I think I got to the bottom of the behaviour, it is described [here ambient vs augmented modules](https://www.typescriptlang.org/docs/handbook/modules/reference.html#ambient-modules). Just adding an export {} is fixing the issue but seems weird to just have it dangling there with no meaning.
> 
> It may look weird but I believe this is the recommended pattern to follow for this situation.
> 
> It has probably worked at some point without `export {}`, but a change in `tsconfig.json` broke it, possible the addition of `"isolatedModules": true`.

The typescript docs state `This module declaration syntax becomes a module augmentation when the file is a module, meaning it has a top-level import or export statement (or is affected by [--moduleDetection force or auto](https://www.typescriptlang.org/tsconfig#moduleDetection)):` and do not indicate any involvement of the `"isolatedModules": true` field.

Honestly I have not seen/found a better option to fix this inconvenience than how i addressed it with the PR #1469

**Comment by atilafassina** [comment]
this should be fixed on the next release, #1469 is merged.

**Comment by bhuynhdev** [comment]
It seems like the doc is not updated yet to use the new `namespace` App. I can write up a PR if the team wanted.

I also hope to add some additional comments on my usage of this feature:
1. I believe the correct usage should be `declare namespace App` instead of `declare module App` (like in the current `todosmvc` example. -> LSP says `A 'namespace' declaration should not be declared using the 'module' keyword. Please use the 'namespace' keyword instead` when I use `declare module App`
2. If one adds an `import` within `global.d.ts` file, the file turns into a module and its type become isolated. As a result, the `App` namespace loses its global-ness, and thus must be written inside a `declare global`. Technically this is a standard Typescript feature, I think it's helpful to document this small "trick" somehow

For example, here is my current `global.d.ts`
```ts
/// <reference types="@solidjs/start/env" />

import type { Request as CfRequest, D1Database, ExecutionContext } from '@cloudflare/workers-types'

declare global {
	namespace App {
		interface RequestEventLocals {
			db: D1Database
		}
	}
}

/**
 * Reference: https://developers.cloudflare.com/workers/runtime-apis/fetch-event/#parameters
 */
export interface CfPagesEnv {
	ASSETS: { fetch: (request: CfRequest) => Promise<Response> }
	CF_PAGES: '1'
	CF_PAGES_BRANCH: string
	CF_PAGES_COMMIT_SHA: string
	CF_PAGES_URL: string

	// Environment variables
	SECRET: string

	// Bindings
	DB: D1Database
}

declare module 'vinxi/http' {
	interface H3EventContext {
		cf: CfRequest['cf']
		cloudflare: {
			request: CfRequest
			env: CfPagesEnv
			context: ExecutionContext
		}
	}
}
```

**Comment by apatrida** [comment]
The answer by @bhuynhdev is correct, the other fixes don't work.  The docs need an updated
